File: reverse-words-in-a-string/solution151.go
// https://leetcode.com/problems/reverse-words-in-a-string/

package main

import (
	"fmt"
	"strings"
)

// T: O(n), S: O(n)
func reverseWords(s string) string {
	arr := strings.Fields(s)
	left, right := 0, len(arr)-1
	for left < right {
		arr[left], arr[right] = arr[right], arr[left]
		left++
		right--
	}
	return strings.Join(arr, " ")
}

func main() {
	s1 := "a good   example"
	s2 := "world hello"
	s3 := "a good   example"
	fmt.Println(reverseWords(s1)) // "example good a"
	fmt.Println(reverseWords(s2)) // "hello world"
	fmt.Println(reverseWords(s3)) // "example good a"
}

File: reverse-words-in-a-string/Solution151.java
// https://leetcode.com/problems/reverse-words-in-a-string/

class Solution151 {

    // T: O(n), S: O(n)
    public String reverseWords(String s) {
        var arr = s.trim().split("\\s+");
        int left = 0, right = arr.length - 1;
        while (left < right) {
            var temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
        return String.join(" ", arr);
    }

    public static void main(String[] args) {
        var sol = new Solution151();
        var s1 = "a good   example";
        var s2 = "world hello";
        var s3 = "a good   example";
        System.out.println(sol.reverseWords(s1)); // "example good a"
        System.out.println(sol.reverseWords(s2)); // "world hello"
        System.out.println(sol.reverseWords(s3)); // "example good a"
    }

}
